In [2]:
library("tidyr")
library("dplyr")

## Data Preprocessing and Analysis
The following libraries are needed to be installed and loaded for the same


In [3]:
print("Hello world")

[1] "Hello world"


In [9]:
data("airquality")
View(airquality)   #here V is capital to view the dataset


Ozone Solar.R Wind Temp Month Day
1    41   190      7.4 67   5      1 
2    36   118      8.0 72   5      2 
3    12   149     12.6 74   5      3 
4    18   313     11.5 62   5      4 
5    NA    NA     14.3 56   5      5 
6    28    NA     14.9 66   5      6 
7    23   299      8.6 65   5      7 
8    19    99     13.8 59   5      8 
9     8    19     20.1 61   5      9 
10   NA   194      8.6 69   5     10 
11    7    NA      6.9 74   5     11 
12   16   256      9.7 69   5     12 
13   11   290      9.2 66   5     13 
14   14   274     10.9 68   5     14 
15   18    65     13.2 58   5     15 
16   14   334     11.5 64   5     16 
17   34   307     12.0 66   5     17 
18    6    78     18.4 57   5     18 
19   30   322     11.5 68   5     19 
20   11    44      9.7 62   5     20 
21    1     8      9.7 59   5     21 
22   11   320     16.6 73   5     22 
23    4    25      9.7 61   5     23 
24   32    92     12.0 61   5     24 
25   NA    66     16.6 57   5     25 
26   NA   266     14.9 58   5     26 
27   NA    NA      8.0 57   5     27 
28   23    13     12.0 67   5     28 
29   45   252     14.9 81   5     29 
30  115   223      5.7 79   5     30 
... ...   ...     ...  ...  ...   ...
124 96    167      6.9 91   9      1 
125 78    197      5.1 92   9      2 
126 73    183      2.8 93   9      3 
127 91    189      4.6 93   9      4 
128 47     95      7.4 87   9      5 
129 32     92     15.5 84   9      6 
130 20    252     10.9 80   9      7 
131 23    220     10.3 78   9      8 
132 21    230     10.9 75   9      9 
133 24    259      9.7 73   9     10 
134 44    236     14.9 81   9     11 
135 21    259     15.5 76   9     12 
136 28    238      6.3 77   9     13 
137  9     24     10.9 71   9     14 
138 13    112     11.5 71   9     15 
139 46    237      6.9 78   9     16 
140 18    224     13.8 67   9     17 
141 13     27     10.3 76   9     18 
142 24    238     10.3 68   9     19 
143 16    201      8.0 82   9     20 
144 13    238     12.6 64   9     21 
145 23     14      9.2 71   9     22 
146 36    139     10.3 81   9     23 
147  7     49     10.3 69   9     24 
148 14     20     16.6 63   9     25 
149 30    193      6.9 70   9     26 
150 NA    145     13.2 77   9     27 
151 14    191     14.3 75   9     28 
152 18    131      8.0 76   9     29 
153 20    223     11.5 68   9     30

ERROR: Error in shape(airquality): could not find function "shape"


In [5]:
str(airquality)  #displays the structure of the dataset

'data.frame':	153 obs. of  6 variables:
 $ Ozone  : int  41 36 12 18 NA 28 23 19 8 NA ...
 $ Solar.R: int  190 118 149 313 NA NA 299 99 19 194 ...
 $ Wind   : num  7.4 8 12.6 11.5 14.3 14.9 8.6 13.8 20.1 8.6 ...
 $ Temp   : int  67 72 74 62 56 66 65 59 61 69 ...
 $ Month  : int  5 5 5 5 5 5 5 5 5 5 ...
 $ Day    : int  1 2 3 4 5 6 7 8 9 10 ...


In [6]:
#drop rows having NA
clean=drop_na(airquality)
View(clean)
str(clean)


Ozone Solar.R Wind Temp Month Day
1    41   190      7.4 67   5      1 
2    36   118      8.0 72   5      2 
3    12   149     12.6 74   5      3 
4    18   313     11.5 62   5      4 
5    23   299      8.6 65   5      7 
6    19    99     13.8 59   5      8 
7     8    19     20.1 61   5      9 
8    16   256      9.7 69   5     12 
9    11   290      9.2 66   5     13 
10   14   274     10.9 68   5     14 
11   18    65     13.2 58   5     15 
12   14   334     11.5 64   5     16 
13   34   307     12.0 66   5     17 
14    6    78     18.4 57   5     18 
15   30   322     11.5 68   5     19 
16   11    44      9.7 62   5     20 
17    1     8      9.7 59   5     21 
18   11   320     16.6 73   5     22 
19    4    25      9.7 61   5     23 
20   32    92     12.0 61   5     24 
21   23    13     12.0 67   5     28 
22   45   252     14.9 81   5     29 
23  115   223      5.7 79   5     30 
24   37   279      7.4 76   5     31 
25   29   127      9.7 82   6      7 
26   71   291     13.8 90   6      9 
27   39   323     11.5 87   6     10 
28   23   148      8.0 82   6     13 
29   21   191     14.9 77   6     16 
30   37   284     20.7 72   6     17 
... ...   ...     ...  ...  ...   ...
82  85    188      6.3 94   8     31 
83  96    167      6.9 91   9      1 
84  78    197      5.1 92   9      2 
85  73    183      2.8 93   9      3 
86  91    189      4.6 93   9      4 
87  47     95      7.4 87   9      5 
88  32     92     15.5 84   9      6 
89  20    252     10.9 80   9      7 
90  23    220     10.3 78   9      8 
91  21    230     10.9 75   9      9 
92  24    259      9.7 73   9     10 
93  44    236     14.9 81   9     11 
94  21    259     15.5 76   9     12 
95  28    238      6.3 77   9     13 
96   9     24     10.9 71   9     14 
97  13    112     11.5 71   9     15 
98  46    237      6.9 78   9     16 
99  18    224     13.8 67   9     17 
100 13     27     10.3 76   9     18 
101 24    238     10.3 68   9     19 
102 16    201      8.0 82   9     20 
103 13    238     12.6 64   9     21 
104 23     14      9.2 71   9     22 
105 36    139     10.3 81   9     23 
106  7     49     10.3 69   9     24 
107 14     20     16.6 63   9     25 
108 30    193      6.9 70   9     26 
109 14    191     14.3 75   9     28 
110 18    131      8.0 76   9     29 
111 20    223     11.5 68   9     30

'data.frame':	111 obs. of  6 variables:
 $ Ozone  : int  41 36 12 18 23 19 8 16 11 14 ...
 $ Solar.R: int  190 118 149 313 299 99 19 256 290 274 ...
 $ Wind   : num  7.4 8 12.6 11.5 8.6 13.8 20.1 9.7 9.2 10.9 ...
 $ Temp   : int  67 72 74 62 65 59 61 69 66 68 ...
 $ Month  : int  5 5 5 5 5 5 5 5 5 5 ...
 $ Day    : int  1 2 3 4 7 8 9 12 13 14 ...


Hence from the above line we can see that first we had 153 rows and now we have 111 rows after dropping the null/NaN values

In [10]:
#chaninging the datatype of the month coloumn
clean$Month=as.factor(clean$Month)
str(clean)

'data.frame':	111 obs. of  6 variables:
 $ Ozone  : int  41 36 12 18 23 19 8 16 11 14 ...
 $ Solar.R: int  190 118 149 313 299 99 19 256 290 274 ...
 $ Wind   : num  7.4 8 12.6 11.5 8.6 13.8 20.1 9.7 9.2 10.9 ...
 $ Temp   : int  67 72 74 62 65 59 61 69 66 68 ...
 $ Month  : Factor w/ 5 levels "5","6","7","8",..: 1 1 1 1 1 1 1 1 1 1 ...
 $ Day    : int  1 2 3 4 7 8 9 12 13 14 ...


In [17]:
#change actual data
newclean=clean %>% mutate(Wind=replace(Wind,match("9",Month) & match("30",Day),12))
View(newclean)
# %>% symbol is called as pipe
#all wind values have been changed to 12
# Not understood this


Ozone Solar.R Wind Temp Month Day
1    41   190     12   67   5      1 
2    36   118     12   72   5      2 
3    12   149     12   74   5      3 
4    18   313     12   62   5      4 
5    23   299     12   65   5      7 
6    19    99     12   59   5      8 
7     8    19     12   61   5      9 
8    16   256     12   69   5     12 
9    11   290     12   66   5     13 
10   14   274     12   68   5     14 
11   18    65     12   58   5     15 
12   14   334     12   64   5     16 
13   34   307     12   66   5     17 
14    6    78     12   57   5     18 
15   30   322     12   68   5     19 
16   11    44     12   62   5     20 
17    1     8     12   59   5     21 
18   11   320     12   73   5     22 
19    4    25     12   61   5     23 
20   32    92     12   61   5     24 
21   23    13     12   67   5     28 
22   45   252     12   81   5     29 
23  115   223     12   79   5     30 
24   37   279     12   76   5     31 
25   29   127     12   82   6      7 
26   71   291     12   90   6      9 
27   39   323     12   87   6     10 
28   23   148     12   82   6     13 
29   21   191     12   77   6     16 
30   37   284     12   72   6     17 
... ...   ...     ...  ...  ...   ...
82  85    188     12   94   8     31 
83  96    167     12   91   9      1 
84  78    197     12   92   9      2 
85  73    183     12   93   9      3 
86  91    189     12   93   9      4 
87  47     95     12   87   9      5 
88  32     92     12   84   9      6 
89  20    252     12   80   9      7 
90  23    220     12   78   9      8 
91  21    230     12   75   9      9 
92  24    259     12   73   9     10 
93  44    236     12   81   9     11 
94  21    259     12   76   9     12 
95  28    238     12   77   9     13 
96   9     24     12   71   9     14 
97  13    112     12   71   9     15 
98  46    237     12   78   9     16 
99  18    224     12   67   9     17 
100 13     27     12   76   9     18 
101 24    238     12   68   9     19 
102 16    201     12   82   9     20 
103 13    238     12   64   9     21 
104 23     14     12   71   9     22 
105 36    139     12   81   9     23 
106  7     49     12   69   9     24 
107 14     20     12   63   9     25 
108 30    193     12   70   9     26 
109 14    191     12   75   9     28 
110 18    131     12   76   9     29 
111 20    223     12   68   9     30

In [25]:
x = mean(clean$Ozone)


In [26]:
#add a new column Quality to the dataset
clean$Quality=as.factor(ifelse(clean$Ozone<x,"Bad","Good"))
View(clean)
count(clean,Quality)
count(clean, Ozone)

Ozone Solar.R Wind Temp Month Day Quality
1    41   190      7.4 67   5      1  Bad    
2    36   118      8.0 72   5      2  Bad    
3    12   149     12.6 74   5      3  Bad    
4    18   313     11.5 62   5      4  Bad    
5    23   299      8.6 65   5      7  Bad    
6    19    99     13.8 59   5      8  Bad    
7     8    19     20.1 61   5      9  Bad    
8    16   256      9.7 69   5     12  Bad    
9    11   290      9.2 66   5     13  Bad    
10   14   274     10.9 68   5     14  Bad    
11   18    65     13.2 58   5     15  Bad    
12   14   334     11.5 64   5     16  Bad    
13   34   307     12.0 66   5     17  Bad    
14    6    78     18.4 57   5     18  Bad    
15   30   322     11.5 68   5     19  Bad    
16   11    44      9.7 62   5     20  Bad    
17    1     8      9.7 59   5     21  Bad    
18   11   320     16.6 73   5     22  Bad    
19    4    25      9.7 61   5     23  Bad    
20   32    92     12.0 61   5     24  Bad    
21   23    13     12.0 67   5     28  Bad    
22   45   252     14.9 81   5     29  Good   
23  115   223      5.7 79   5     30  Good   
24   37   279      7.4 76   5     31  Bad    
25   29   127      9.7 82   6      7  Bad    
26   71   291     13.8 90   6      9  Good   
27   39   323     11.5 87   6     10  Bad    
28   23   148      8.0 82   6     13  Bad    
29   21   191     14.9 77   6     16  Bad    
30   37   284     20.7 72   6     17  Bad    
... ...   ...     ...  ...  ...   ... ...    
82  85    188      6.3 94   8     31  Good   
83  96    167      6.9 91   9      1  Good   
84  78    197      5.1 92   9      2  Good   
85  73    183      2.8 93   9      3  Good   
86  91    189      4.6 93   9      4  Good   
87  47     95      7.4 87   9      5  Good   
88  32     92     15.5 84   9      6  Bad    
89  20    252     10.9 80   9      7  Bad    
90  23    220     10.3 78   9      8  Bad    
91  21    230     10.9 75   9      9  Bad    
92  24    259      9.7 73   9     10  Bad    
93  44    236     14.9 81   9     11  Good   
94  21    259     15.5 76   9     12  Bad    
95  28    238      6.3 77   9     13  Bad    
96   9     24     10.9 71   9     14  Bad    
97  13    112     11.5 71   9     15  Bad    
98  46    237      6.9 78   9     16  Good   
99  18    224     13.8 67   9     17  Bad    
100 13     27     10.3 76   9     18  Bad    
101 24    238     10.3 68   9     19  Bad    
102 16    201      8.0 82   9     20  Bad    
103 13    238     12.6 64   9     21  Bad    
104 23     14      9.2 71   9     22  Bad    
105 36    139     10.3 81   9     23  Bad    
106  7     49     10.3 69   9     24  Bad    
107 14     20     16.6 63   9     25  Bad    
108 30    193      6.9 70   9     26  Bad    
109 14    191     14.3 75   9     28  Bad    
110 18    131      8.0 76   9     29  Bad    
111 20    223     11.5 68   9     30  Bad

Quality n 
1 Bad     69
2 Good    42

Ozone n  
1    1    1  
2    4    1  
3    6    1  
4    7    2  
5    8    1  
6    9    3  
7   10    1  
8   11    3  
9   12    2  
10  13    4  
11  14    4  
12  16    4  
13  18    4  
14  19    1  
15  20    4  
16  21    4  
17  22    1  
18  23    6  
19  24    2  
20  27    1  
21  28    2  
22  29    1  
23  30    2  
24  31    1  
25  32    3  
26  34    1  
27  35    1  
28  36    2  
29  37    2  
30  39    2  
... ...   ...
37   48   1  
38   49   1  
39   50   1  
40   52   1  
41   59   2  
42   61   1  
43   63   1  
44   64   2  
45   65   1  
46   71   1  
47   73   2  
48   76   1  
49   77   1  
50   78   1  
51   79   1  
52   80   1  
53   82   1  
54   84   1  
55   85   2  
56   89   1  
57   91   1  
58   96   1  
59   97   2  
60  108   1  
61  110   1  
62  115   1  
63  118   1  
64  122   1  
65  135   1  
66  168   1

In [27]:
#calculating mean of an array or given data
y=mean(c(0,3,4,7,1))
print(y)

#median
y=median(clean$Wind)
print(y)

[1] 3
[1] 9.7


In [32]:
#to calculate mode, R does not provide a standard function

#match returns the position of the given element
match(2,c(8,3,4,2,1))


[1] 4

In [33]:
#calculating mode
getmode=function(v){
  uniqv=unique(v)
  uniqv[which.max(tabulate(match(v,uniqv)))]
}

In [34]:
count(clean,Quality)
modeQuality=getmode(clean$Quality)
print(modeQuality)


Quality n 
1 Bad     69
2 Good    42

[1] Bad
Levels: Bad Good


In [35]:
modeOzone=getmode(clean$Ozone)
print(modeOzone)

[1] 23
